### PR TITLE
Rate limit [DPP-999]

### PR DESCRIPTION
**Pull request by simonmaxen-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala [hunk 1]
```diff
@@ -48,6 +48,14 @@ final case class UserManagementConfig(
     maxUsersPageSize: Int,
 )
 
+final case class RateLimitingConfig(
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala:51]: These classes would be  more at home somewhere nearer to the rate limiting mechanism itself. User management is rather unrelated.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    } else {
+      next.startCall(call, headers)
+    }
+
+  }
+
+}
+
+object RateLimitingInterceptor {
+  val serverReflectionInfo = "grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo"
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:56]: Health service needs to be excluded as well

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import io.grpc.Status.Code
+import io.grpc._
+import io.grpc.protobuf.StatusProto
+
+private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config: RateLimitingConfig)
+    extends ServerInterceptor {
+
+  import metrics.daml.lapi.threadpool.apiServices
+
+  /** Match naming in [[com.codahale.metrics.InstrumentedExecutorService]] */
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:19]: (note that if the metric was created with a different metric type then an error results)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    val fullMethodName = call.getMethodDescriptor.getFullMethodName
+
+    if (queued > config.maxApiServicesQueueSize && (fullMethodName != serverReflectionInfo)) {
+      val rpcStatus = com.google.rpc.Status
+        .newBuilder()
+        .setCode(Code.ABORTED.value())
+        .setMessage(
+          s"The api services queue size ($queued) has exceeded the maximum (${config.maxApiServicesQueueSize}).  Api services metrics are available at $apiServices"
+        )
+        .build()
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:41]: The plan is to use self-service error codes in  a follow up PR.  See DPP-1000 for details.

**Review by rautenrieth-da — APPROVED** [review]

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    withChannel(metrics, new HelloServiceAkkaImplementation, config).use { channel: Channel =>
+      val helloService = HelloServiceGrpc.stub(channel)
+      val submitted = metrics.registry.meter(
+        MetricRegistry.name(metrics.daml.lapi.threadpool.apiServices, "submitted")
+      )
+      for {
+        _ <- helloService.single(HelloRequest(1))
+        _ = submitted.mark(config.maxApiServicesQueueSize.toLong + 1)
+        exception <- helloService.single(HelloRequest(2)).failed
+        _ = submitted.mark(-config.maxApiServicesQueueSize.toLong)
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:64]: Done.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    val queued = submitted.getCount - running.getCount - completed.getCount
+
+    val fullMethodName = call.getMethodDescriptor.getFullMethodName
+
+    if (queued > config.maxApiServicesQueueSize && (fullMethodName != serverReflectionInfo)) {
+      val rpcStatus = com.google.rpc.Status
+        .newBuilder()
+        .setCode(Code.ABORTED.value())
+        .setMessage(
+          s"The api services queue size ($queued) has exceeded the maximum (${config.maxApiServicesQueueSize}).  Api services metrics are available at $apiServices"
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala]: We can fine tune this in DPP-1000.  However a decision was made that we should inform the user of the metric driving the decision to help with problem diagnosis.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      call: ServerCall[ReqT, RespT],
+      headers: Metadata,
+      next: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+
+    val queued = submitted.getCount - running.getCount - completed.getCount
+
+    val fullMethodName = call.getMethodDescriptor.getFullMethodName
+
+    if (queued > config.maxApiServicesQueueSize && (fullMethodName != serverReflectionInfo)) {
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala]: We are moving in that direction as @mziolekda has already asked that the health endpoint is not limited (this `doNotLimit` whitelist can be extended as needed).

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      next: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+
+    val queued = submitted.getCount - running.getCount - completed.getCount
+
+    val fullMethodName = call.getMethodDescriptor.getFullMethodName
+
+    if (queued > config.maxApiServicesQueueSize && !doNonLimit.contains(fullMethodName)) {
+
+      val rpcStatus = com.google.rpc.Status
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:39]: Should we use an error code? (from //ledger/error)

**Pull request merged**